orc8r/cloud/go/services/certifier/servicers/certifier.go

```diff
@@ -712,60 +718,166 @@ func (srv *CertifierServer) getPolicyDecisionFromTokenMany(tokens *certprotos.To
+// Returns false otherwise
+func doResourceTypesMatch(req *certprotos.Request, policy *certprotos.Policy) bool {
+	reqType := getRequestType(req)
+	policyType := getPolicyType(policy)
+	return req.GetResourceId() == nil || policyType == constants.Path || reqType == policyType
+}
+
+// isActionAuthorized checks if the requested action to read/write is authorized by the policy
+// Returns the effect if the policy allows/denies write access regardless of requested action,
+// or if policy allows/denies read access and the requested action is read
+// Otherwise, returns an unknown effect, which implies that the policy does not apply to this requested resource
+func isActionAuthorized(req *certprotos.Request, policy *certprotos.Policy) certprotos.Effect {
 	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
+	if !doResourceTypesMatch(req, policy) {
 		return certprotos.Effect_UNKNOWN
 	}
-	reqPath := reqResource.Resource
-	policyPath := policyResource.Resource
-	if ok, _ := doublestar.Match(policyPath, reqPath); ok {
-		return policyResource.Effect
+	if policy.Action == certprotos.Action_WRITE {
+		return policy.Effect
+	}
+	if policy.Action == certprotos.Action_READ && req.Action == certprotos.Action_READ {
+		return policy.Effect
 	}
 	return certprotos.Effect_UNKNOWN
 }
 
-// checkAction checks if the requested action is authorized by the policy
-func checkAction(reqResource *certprotos.Resource, policyResource *certprotos.Resource) certprotos.Effect {
+// isResourceAuthorized checks if the requested resource (either a path, network, or tenant) is authorized by the policy
+// Returns the effect if the policy's path matches the requested path and if the policy's tenant/network ID matches that
+// of the requested when applicable
+// Otherwise, returns an unknown effect, which implies that the policy does not apply to this requested resource
+func isResourceAuthorized(req *certprotos.Request, policy *certprotos.Policy) certprotos.Effect {
 	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
+	if !doResourceTypesMatch(req, policy) {
 		return certprotos.Effect_UNKNOWN
 	}
-	if policyResource.Action == certprotos.Action_WRITE {
-		return policyResource.Effect
+
+	finalEffect := certprotos.Effect_UNKNOWN
+	if path := policy.GetPath(); path != nil {
+		if ok, _ := doublestar.Match(path.Path, req.Resource); ok {
+			finalEffect = policy.Effect
+		}
 	}
-	if policyResource.Action == certprotos.Action_READ && reqResource.Action == certprotos.Action_READ {
-		return policyResource.Effect
+
+	reqType := getRequestType(req)
+	policyType := getPolicyType(policy)
+	if reqType == constants.NetworkID && policyType == constants.NetworkID {
+		reqID := req.GetNetworkId()
+		policyIDs := policy.GetNetwork().Networks
+		for _, policyID := range policyIDs {
+			if finalEffect != certprotos.Effect_DENY && reqID == policyID {
```

Comment:
Rather than checking if == deny on every loop, just return early when you hit a deny